specifyweb/frontend/js_src/lib/components/Statistics/hooks.tsx

```diff
@@ -0,0 +1,424 @@
+      headers: {
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        Accept: 'application/json',
+      },
+      body: keysToLowerCase({
+        ...serializeResource(query),
+        countOnly: true,
+      }),
+    }).then(({ data }) => formatNumber(data.count));
+    return ajaxPromise;
+  };
+}
+
+export const useQuerySpecToResource = (
+  label: string,
+  tableName: keyof Tables,
+  fields: RA<
+    Partial<SerializedResource<SpQueryField>> & { readonly path: string }
+  >
+): SpecifyResource<SpQuery> =>
+  React.useMemo(
+    () =>
+      deserializeResource(
+        addMissingFields('SpQuery', {
+          name: label,
+          contextName: tableName,
+          contextTableId: schema.models[tableName].tableId,
+          countOnly: false,
+          selectDistinct: false,
+          fields: fields.map(({ path, ...field }, index) =>
+            serializeResource(
+              makeQueryField(tableName, path, {
+                ...field,
+                position: index,
+              })
+            )
+          ),
+        })
+      ),
+    [label, tableName, fields]
+  );
+
+export function useResolveStatSpec(
+  item: CustomStat | DefaultStat,
+  statsSpec: StatsSpec
+): StatSpecCalculated {
+  return React.useMemo(() => {
+    if (item.type === 'CustomStat') {
+      return {
+        type: 'QueryStat',
+        tableName: item.tableName,
+        fields: item.fields,
+      };
+    } else {
+      const statSpecItem =
+        statsSpec[item.pageName][item.categoryName].items[item.itemName];
+      return item.itemType === 'BackendStat'
+        ? {
+            type: 'BackEndStat',
+            pathToValue:
+              item.pathToValue ??
+              (statSpecItem.spec as BackEndStat).pathToValue,
+            urlToFetch: (statSpecItem.spec as BackEndStat).urlToFetch,
+            formatter: (statSpecItem.spec as BackEndStat).formatter,
+          }
+        : {
+            type: 'QueryStat',
+            tableName: (statSpecItem.spec as QueryBuilderStat).tableName,
+            fields: (statSpecItem.spec as QueryBuilderStat).fields,
+          };
+    }
+  }, [item]);
+}
+
+export function useCategoryToFetch(
+  layout: StatLayout | undefined
+): WritableArray<string> {
+  return React.useMemo(() => {
+    if (layout === undefined) return [];
+    const categoriesToFetch: WritableArray<string> = [];
+    layout.forEach((pageLayout) =>
+      pageLayout.categories.forEach(({ items, categoryToFetch }) => {
+        if (items === undefined && categoryToFetch !== undefined) {
+          categoriesToFetch.push(categoryToFetch);
+        }
+        (items ?? []).forEach((item) => {
+          if (
+            item.type === 'DefaultStat' &&
+            item.itemType === 'BackendStat' &&
+            item.itemValue === undefined
+          )
+            categoriesToFetch.push(item.categoryName);
+        });
+      })
+    );
+    return categoriesToFetch.filter((categoryToFetch) =>
```

Comment:
categoriesToFetch is all the unknown categories that need to be fetched. Since I already have a function that returns the previous promise if category name is same (the throttledAjax handles it), so it doesn't matter if categories contain duplicates.